File: src/content/docs/topics/cs/coding-problems/linked-list/024-swap-nodes-in-pairs.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func swapPairs(head *ListNode) *ListNode {
	// TODO: implement
	return nil
}

func makeList(vals []int) *ListNode {
	if len(vals) == 0 {
		return nil
	}
	head := &ListNode{Val: vals[0]}
	cur := head
	for _, v := range vals[1:] {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return head
}

func listToSlice(head *ListNode) []int {
	var result []int
	for head != nil {
		result = append(result, head.Val)
		head = head.Next
	}
	return result
}

func slicesEqual(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func runTests() {
	assert(slicesEqual(listToSlice(swapPairs(makeList([]int{1, 2, 3, 4}))), []int{2, 1, 4, 3}))
	assert(slicesEqual(listToSlice(swapPairs(nil)), []int{}))
	assert(slicesEqual(listToSlice(swapPairs(makeList([]int{1}))), []int{1}))
	assert(slicesEqual(listToSlice(swapPairs(makeList([]int{1, 2, 3}))), []int{2, 1, 3}))
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/linked-list/024-swap-nodes-in-pairs.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val: number = 0, next: ListNode | null = null) {
        this.val = val; this.next = next;
    }
}

function toList(head: ListNode | null): number[] {
    const out: number[] = [];
    while (head) { out.push(head.val); head = head.next; }
    return out;
}

function fromList(vals: number[]): ListNode | null {
    const dummy = new ListNode();
    let cur = dummy;
    for (const v of vals) { cur.next = new ListNode(v); cur = cur.next; }
    return dummy.next;
}

function swapPairs(head: ListNode | null): ListNode | null {
    // TODO: implement
    return null;
}

function _runTests(): void {
    assert(JSON.stringify(toList(swapPairs(fromList([1,2,3,4])))) === JSON.stringify([2,1,4,3]));
    assert(JSON.stringify(toList(swapPairs(fromList([])))) === JSON.stringify([]));
    assert(JSON.stringify(toList(swapPairs(fromList([1])))) === JSON.stringify([1]));
    assert(JSON.stringify(toList(swapPairs(fromList([1,2,3])))) === JSON.stringify([2,1,3]));
    // perf
    const head = fromList(Array.from({ length: 1000 }, (_, i) => i));
    const t0 = performance.now();
    swapPairs(head);
    console.log(`perf swapPairs(1000 nodes): ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
